server/util.go

```diff
@@ -110,8 +136,8 @@ func SetCmdServerContext(cmd *cobra.Command, serverCtx *Context) error {
 // configuration file. The Tendermint configuration file is parsed given a root
 // Viper object, whereas the application is parsed with the private package-aware
 // viperCfg object.
-func interceptConfigs(ctx *Context, rootViper *viper.Viper) (*tmcfg.Config, error) {
-	rootDir := rootViper.GetString(flags.FlagHome)
+func interceptConfigs(tmViper *viper.Viper, cosmosViper *viper.Viper) (*tmcfg.Config, error) {
```

Comment:
What? Why are we doing this?

Reply:
Well if you look at the master branch `app.toml` and `config.toml` are separate files. Logically, they need to become separate configuration instances. Otherwise values from one could overwrite the other. I added comments to `InterceptConfigsPreRunHandler` to try and make it clearer.

Reply:
No, they should be the same instance. This was intentional. Also, no two configs could overwrite one another as the configurations have completely different contexts. If there are conflicting keys, then that's a bug.

Reply:
I'm confused by your statement, in master two separate instances are created.

Reply:
1. Creation of the first viper instance here https://github.com/cosmos/cosmos-sdk/blob/9856327d95eba4c3ae2d15d26a76ad3ac17b20b8/server/util.go#L58
2. Creation of the second viper instance here https://github.com/cosmos/cosmos-sdk/blob/9856327d95eba4c3ae2d15d26a76ad3ac17b20b8/server/util.go#L43
3. Use for first viper instance is here https://github.com/cosmos/cosmos-sdk/blob/9856327d95eba4c3ae2d15d26a76ad3ac17b20b8/server/util.go#L114
4. Use of the second viper instance is here  https://github.com/cosmos/cosmos-sdk/blob/9856327d95eba4c3ae2d15d26a76ad3ac17b20b8/server/util.go#L157 and in other places in `server/start.go`

Reply:
Yes, we should be using a single viper instance, which originally your PR addressed if I'm not mistaken.

Reply:
Well, it initially makes sense to me to use one instance. But with them being separate files it makes sense to have them be separate configuration objects. For example consider this scenario with a single viper instance

1. Someone defines a `config.toml` with all values normally found in `config.toml` & `app.toml`
2. `app.toml` is an empty file
3. The `start` command is run
4. First the code reads in `config.toml` & sets up tendermint configuration
5. Next the code reads in `app.toml` (an empty file, so no change) & sets up the configuration for this application.

This is possible because viper instances have an internal state (for example `Viper.config`) that gets updated when functions like `ReadConfig` are called. `ReadConfig` is delegating to functions like `yaml.Unmarshal` and `toml.LoadReader` which do not clear the map of other keys as far as I know. So each call to `ReadConfig` is additive. For example: https://github.com/spf13/viper/blob/master/viper.go#L1608

In my above example the configuration for the `start` command could get pulled only from the `config.toml` file. So it would work but be confusing to an end user. I would not expect a user to do this, but it could happen by accident.

In other words: why have two configuration files if internally they will get merged into a single Viper instance anyways?

Reply:
> Someone defines a config.toml with all values normally found in config.toml & app.toml

This is not a situation that should happen. The set of configurations found in `config.toml` and `app.toml` should be entirely unique and separate. Hence, we shouldn't need to worry about merge conflicts.

Reply:
OK I'll write a unit test to cover that then.

Reply:
Awesome! In addition, having to handle two config objects would not be ideal I think.